# Practice case using pymongo

Pada case ini akan dilakukan analysis menggunakan pymongo. Pymongo adalah salah satu packages di python yang digunakan untuk connect ke MongoDB. MongoDB adalah salah satu platfrom untuk NoSQL program.

**Problem**:
- Membuat collection baru dengan nama clean_movies yang sama persis dengan movies collection pada database sample_mflix dengan memakai collection movies_intial
- Validasi collection clean_movies
    - Semua document pada clean_movies dan movie sama
    - Banyak document pada clean_movies dan movie sama 
    - Semua fields pada clean_movies ada pada movie 
    - Semua value pada clean_movies sama dengan semua value pada movies dengan urutan yang sama
    
**Outline**:
<ul>
<li><a href="#intro">Introduction</a></li>
<li><a href="#collection">1. Create New Collection</a></li>
<li><a href="#validation">2. Validation </a></li>
<li><a href="#conclusions">Conclusions</a></li>
</ul>

<a id='intro'></a>
## Introduction

**Import Libraries**

In [1]:
#Import Libraries
#For pymongo
import pymongo
from pymongo import MongoClient
import pprint
from IPython.display import clear_output

#For datetime converting
from datetime import datetime

### Showing 'movies_initial' Collection

**Import Cluster for Collection 'movie_initial'** 

In [2]:
#Import cluster that contain movies_initial collection
client = MongoClient("mongodb+srv://admin1234:12345@cluster0-miqju.gcp.mongodb.net/test?retryWrites=true&w=majority")
print(client)

MongoClient(host=['cluster0-shard-00-00-miqju.gcp.mongodb.net:27017', 'cluster0-shard-00-01-miqju.gcp.mongodb.net:27017', 'cluster0-shard-00-02-miqju.gcp.mongodb.net:27017'], document_class=dict, tz_aware=False, connect=True, retrywrites=True, w='majority', authsource='admin', replicaset='Cluster0-shard-0', ssl=True)


**Showing Database Names in Cluster**

In [3]:
#show all database in cluster
print(client.list_database_names())

['sample_mflix', 'admin', 'local']


Dari hasil diatas didapatkan informasi nama database yang berada di cluster yang digunakan yang telah di definisikan di client. Terlihat bahwa terdapat 3 database di client yaitu 'sample_mflix', 'admin', 'local'. 

Untuk analisis selanjutnya akan digunakan database 'sample_mflix' yang didalamnya terdapat collection 'movies_initial'.

**Showing Collection in Database 'sample_mflix'**

In [4]:
db = client['sample_mflix']
col = db.movies_initial

# show all collecton in used database
print(db.list_collection_names())

['clean_movies_Putrisqiana', 'clean_movies_tara', 'clean_movies_bunga', 'clean_movies_zumar', 'movies_initial', 'clean_movies_putrisqiana', 'clean_movies_firdaus', 'clean_movies_dwilaras', 'clean_movies_nurlailiis', 'clean_movies_faizah']


Hasil diatas memuat informasi nama-nama collection(table) yang berada di database 'sample_mflix'.

**Showing one document(row) with randomly in collection 'movies_initial'** 

In [5]:
pprint.pprint(col.find_one())

{'_id': ObjectId('5daffb1b986c4adedb754af9'),
 'awards': '',
 'cast': 'Carmencita',
 'country': 'USA',
 'director': 'William K.L. Dickson',
 'fullplot': 'Performing on what looks like a small wooden stage, wearing a '
             'dress with a hoop skirt and white high-heeled pumps, Carmencita '
             'does a dance with kicks and twirls, a smile always on her face.',
 'genre': 'Documentary, Short',
 'imdbID': 1,
 'imdbRating': 5.9,
 'imdbVotes': 1032,
 'language': '',
 'lastupdated': '2015-08-26 00:03:45.040000000',
 'metacritic': '',
 'plot': 'Performing on what looks like a small wooden stage, wearing a dress '
         'with a hoop skirt and white high-heeled pumps, Carmencita does a '
         'dance with kicks and twirls, a smile always on her face.',
 'poster': 'http://ia.media-imdb.com/images/M/MV5BMjAzNDEwMzk3OV5BMl5BanBnXkFtZTcwOTk4OTM5Ng@@._V1_SX300.jpg',
 'rating': 'NOT RATED',
 'released': '',
 'runtime': '1 min',
 'title': 'Carmencita',
 'type': 'movie',
 'writer':

### Showing 'movie' collection

Karena cluster yang digunakan untuk 'movie_initial' tidak memuat collection(table) 'movies' maka selanjutnya untuk menampilkan collection 'movies' digunakan cluster yang berbeda yang nantinya diberi nama 'client2'.

**Get Cluster that Contains 'Movies'**

In [6]:
client2 = MongoClient("mongodb+srv://userstudent:admin1234@cluster0-nnbxe.gcp.mongodb.net/test?retryWrites=true& w=majority")

C:\Users\faizah064287\AppData\Local\Continuum\anaconda3\lib\site-packages\pymongo\common.py:756: UserWarning: Unknown option  w
  warnings.warn(str(exc))


**Initialization Collection that Used**

In [7]:
dbmovies=client2['sample_mflix']
colmovies = dbmovies.movies

**Showing one document(row) with randomly in collection 'movies'** 

In [8]:
pprint.pprint(colmovies.find_one())

{'_id': ObjectId('573a1390f29313caabcd5a93'),
 'awards': {'nominations': 0, 'text': '1 win.', 'wins': 1},
 'cast': ['Howard C. Hickman', 'Enid Markey', 'Lola May', 'Kate Bruce'],
 'countries': ['USA'],
 'directors': ['Reginald Barker',
               'Thomas H. Ince',
               'Raymond B. West',
               'Walter Edwards',
               'David Hartford',
               'Jay Hunt',
               'J. Parker Read Jr.'],
 'fullplot': 'Allegorical film about peace. A king starts a war, many of the '
             'women are against it, people are pressed into service. A count '
             'has constructed a submarine and gets the order to sink an ocean '
             'liner, that is also carrying - supposedly - ammunition for the '
             'enemy. The count refuses to fire the torpedos, and sinks the '
             'submarine. He survives, but in a limbo between death and life '
             'where he meets Jesus, who takes him over to preach peace. '
             'Natura

**Get Unique Keys in 'movies' Collection**

Akan ditampilkan **unique keys** pada **movies** collection untuk memperoleh informasi tentang field/keys. Informasi ini akan digunakan untuk projecting pembentukan **clean_movies_faizah** dimana collection tersebut harus memiliki field yang sama dengan **movies** collection.

In [20]:
a = [list(doc.keys()) for doc in colmovies.find({})]
a = set([j for i in a for j in i])
a

{'_id',
 'awards',
 'cast',
 'countries',
 'directors',
 'fullplot',
 'genres',
 'imdb',
 'languages',
 'lastupdated',
 'metacritic',
 'num_mflix_comments',
 'plot',
 'poster',
 'rated',
 'released',
 'runtime',
 'title',
 'tomatoes',
 'type',
 'writers',
 'year'}

Diatas adalah hasil **unique keys** pada **movies** collection.

<a id='collection'></a>
## 1. Create New Collection

**Identify 'IMDB' keys that used in the 'movie' collection**

Karena dalam pembentukan collection baru, document pada collection baru harus sama dengan collection 'movies' maka dilakukan identifikasi field/keys **imdb** pada collection 'movies'. Sehingga, nantinya document pada collection baru sama persis dengan collection 'movies' dimana diambil dari colllection 'movies_initial'. 

In [9]:
imdb = list(colmovies.find({},{'_id': 0, 'imdb':1}))
imdb_list = []
print(len(imdb))
for tit in imdb:
    imdb_list.append(list(tit.values())[0])
imdb_list[:5]

23541


[{'id': 6517, 'rating': 6.3, 'votes': 162},
 {'id': 8133, 'rating': 7.8, 'votes': 4680},
 {'id': 12512, 'rating': 6.8, 'votes': 489},
 {'id': 13025, 'rating': 7.8, 'votes': 3738},
 {'id': 15498, 'rating': 7.1, 'votes': 327}]

Dari hasil diatas, maka dari hasil diatas akan digunakan untuk matching data ke **movies_initial** untuk menghasilkan **document** yang sama dengan **movies**. Matching document ini menggunakan fungsi **match**.

**Create new collection 'clean_movies_faizah' using projecting**

In [68]:
pipeline = [
    {
        ### This code is to include the keys for a new collection on database 
        ### 1 for include , 0 for exclude the keys 
        ### Split some keys into an array value, renaming as well
        ### Renaming the keys
        ### grouping the several keys into embeded documents by one keys 
        '$project': {
            'awards': {'nominations': {'$arrayElemAt':[{'$split':[{'$arrayElemAt':[{'$split': [ "$awards", " nomination" ]}, -2]},' ']},-1]},
                       'text': "$awards",
                       'wins': {'$arrayElemAt':[{'$split':[{'$arrayElemAt':[{'$split': [ "$awards", " win" ]}, -2]},' ']},-1]}
                      },
            'cast': {'$split': ["$cast", ", "]},
            'countries': {'$split': ["$country", ", "]},
            'directors': {'$split': ["$director", ", "]},
            'fullPlot':  {"$cond": [{"$eq": ["$fullplot", ""]}, "$nonExistinField", "$fullplot"]},
            'genres': {'$split': ["$genre", ", "]},
            'imdb': {
                'id': "$imdbID",
                'rating': "$imdbRating",
                'votes': "$imdbVotes"
                },
            'languages': {"$cond": [{"$eq": ["$language", ""]}, "$nonExistinField", "$language"]},
            'lastupdated':  {"$cond": [{"$eq": ["$lastupdated", ""]}, "$nonExistinField", "$lastupdate"]},
            'metacritic':{"$cond": [{"$eq": ["$metacritic", ""]}, "$nonExistinField", "$metacritic"]},
            'num_mflix_comments':{"$sum" : 1},
            'rated' : {"$cond": [{"$eq": ["$rating", ""]}, "$nonExistinField", "$rating"]},
            'plot': {"$cond": [{"$eq": ["$plot", ""]}, "$nonExistinField", "$plot"]},
            'poster': {"$cond": [{"$eq": ["$poster", ""]}, "$nonExistinField", "$poster"]},
            'released': { '$concat': [ "datetime.datetime(", 
                                      {'$arrayElemAt':[{'$split': [ "$released", "-" ]}, 0]},
                                      ", ",
                                      {'$arrayElemAt':[{'$split': [ "$released", "-" ]}, 1]},
                                      ", ",
                                      {'$arrayElemAt':[{'$split': [ "$released", "-" ]}, 2]},
                                      ", 0, 0)" ] },
            'runtime': {"$cond": [{'$eq': ["$runtime", ""]}, 0 ,{'$arrayElemAt':[{'$split': [ "$runtime", " min" ]}, 0]}]},
            'title': {"$cond": [{"$eq": ["title", ""]}, "$nonExistinField", "$title"]},
            'tomatoes' :{
                'lastUpdated': "",
                'viewer' : {'numReviews' : "", 'rating' : ""}
            },
            'type': {"$cond": [{"$eq": ["$type", ""]}, "$nonExistinField", "$type"]},
            'writers': {'$split': ["$writer", ", "]},
            'year': {"$cond": [{"$eq": ["$year", ""]}, "$nonExistinField", "$year"]}
        }
    },{
        "$match" : {"imdb" : {"$in" : [dict(sorted(imdb_list[x].items(), key=lambda x: x[0])) for x in range(len(imdb_list))]}}
      },
    {
        ### naming the new collection as output ### 
        '$out': "clean_movies_faizah"
    }
]

clear_output()

##Information about collection that used to fill 'clean_movies_faizah'
pprint.pprint(list(client.sample_mflix.movies_initial.aggregate(pipeline)))

[]


Informasi pada pembentukan document dengan projecting:
>- split : artinya memisahkan nilai pada keys tersebut yang dipisahkan oleh koma(,) dimana terdapat beberapa keys yang memiliki values lebih dari satu.
>- cond : menampilkan keys tersebut jika keys tersebut memiliki values dan tidak akan menampilkan keys tersebut jika keys tersebut tidak memiliki values.

>Sebagian besar saya mengambil keys dari document menggunakan **cond** karena setelah diperhatikan pada collection **movies_initial** ada beberapa keys yang ada values/isinya dan tidak ada values/isinya, sehingga jika hanya menampilkan akan ada beberapa keys yang ditampilkan tetapi memiliki values kosong, padahal pada format collection **movie** hanya akan menampilkan keys yang berisi values.

>- embeded : menggabungkan beberapa keys menjadi satu keys yang merepresentasikan kumpulan keys tersebut.
>- concat : menggabungkan 2 values.
>- renaming : mengganti nama keys.
>- contoh :'id': imdbID, mengganti nama 'id' dimana berisi keys : imdbID.
>- sum : 1 : return hasilnya adalah 1.
>- aggregate : mengisi collection baru dengan collection **movies_initial** pada database **sample_mflix**.

**Initialization clean_movies_faizah collection**

In [69]:
dbmovies = client['sample_mflix']
colmoviesi = dbmovies.clean_movies_faizah

**Showing one document(rows) with randomly in clean_movies_faizah**

In [70]:
pprint.pprint(colmoviesi.find_one())

{'_id': ObjectId('5daffb1b986c4adedb754afa'),
 'awards': {'nominations': None, 'text': '1 win.', 'wins': '1'},
 'cast': ['Charles Kayser', 'John Ott'],
 'countries': ['USA'],
 'directors': ['William K.L. Dickson'],
 'fullPlot': 'A stationary camera looks at a large anvil with a blacksmith '
             'behind it and one on either side. The smith in the middle draws '
             'a heated metal rod from the fire, places it on the anvil, and '
             'all three begin a rhythmic hammering. After several blows, the '
             'metal goes back in the fire. One smith pulls out a bottle of '
             'beer, and they each take a swig. Then, out comes the glowing '
             'metal and the hammering resumes.',
 'genres': ['Short'],
 'imdb': {'id': 5, 'rating': 6.2, 'votes': 1189},
 'num_mflix_comments': 1,
 'plot': 'Three men hammer on an anvil and pass a bottle of beer around.',
 'rated': 'UNRATED',
 'released': 'datetime.datetime(1893, 05, 09, 0, 0)',
 'runtime': '1',
 't

<a id='validation'></a>
## 2. Validation

### 1. Get length of movies and clean_movie_faizah

length document of **clean_movies_faizah** collection

In [14]:
len(list(colmoviesi.find()))

23539

length document of **movies** collection

In [15]:
len(list(colmovies.find()))

23541

Dari hasil diatas terdapat informasi tentang jumlah document(row) di collection:
- Untuk collection 'clean_movies_faizah' terdapat 25339 document
- Untuk collection 'movies' terdapat 23541 document

Maka, dari itu, terdapat 2 document yang berbeda. Maka, akan dicari manakah dokumen yang berbeda tersebut.

**Showing document that different between 'clean_movies_faizah' and 'movies'**

Pada case ini akan ditampilkan **id** yang berbeda dari dua collection tersebut.

In [71]:
grup = list(colmoviesi.aggregate([{"$sortByCount":"$imdb"}]))
grup2 = list(colmovies.aggregate([{"$sortByCount":"$imdb"}]))

for i in grup2:
  if i not in grup:
    print(i)

{'_id': {'id': 5, 'rating': 6.2, 'votes': 1189}, 'count': 2}


Dari hasil diatas, terdapat informasi tentang **id** yang berbeda, terlihat bahwa ada document yang redundant di **movies** collection ini terlihat karena 'count' = 2. 
Sehingga, disimpulkan bahwa **movies** collection belum bersih artinya masih ada data redundant, sedangkan **clean_movies_faizah** sudah tidak ada data redundant.

In [72]:
#Showing 'id:5' in clean_movies_faizah collection
list(colmoviesi.find({'imdb':{'id': 5, 'rating': 6.2, 'votes': 1189}}))

[{'_id': ObjectId('5daffb1b986c4adedb754afa'),
  'awards': {'nominations': None, 'text': '1 win.', 'wins': '1'},
  'cast': ['Charles Kayser', 'John Ott'],
  'countries': ['USA'],
  'directors': ['William K.L. Dickson'],
  'fullPlot': 'A stationary camera looks at a large anvil with a blacksmith behind it and one on either side. The smith in the middle draws a heated metal rod from the fire, places it on the anvil, and all three begin a rhythmic hammering. After several blows, the metal goes back in the fire. One smith pulls out a bottle of beer, and they each take a swig. Then, out comes the glowing metal and the hammering resumes.',
  'genres': ['Short'],
  'imdb': {'id': 5, 'rating': 6.2, 'votes': 1189},
  'num_mflix_comments': 1,
  'plot': 'Three men hammer on an anvil and pass a bottle of beer around.',
  'rated': 'UNRATED',
  'released': 'datetime.datetime(1893, 05, 09, 0, 0)',
  'runtime': '1',
  'title': 'Blacksmith Scene',
  'tomatoes': {'lastUpdated': '', 'viewer': {'numReview

In [73]:
#Showing 'id:5' in movies collection
list(colmovies.find({'imdb':{'id': 5, 'rating': 6.2, 'votes': 1189}}))

[{'_id': ObjectId('5dafd0d3d75e5fa8b23dae21'),
  'awards': {'nominations': 0, 'text': '1 win.', 'wins': 1},
  'cast': ['Charles Kayser', 'John Ott'],
  'countries': ['USA'],
  'directors': ['William K.L. Dickson'],
  'fullplot': 'A stationary camera looks at a large anvil with a blacksmith behind it and one on either side. The smith in the middle draws a heated metal rod from the fire, places it on the anvil, and all three begin a rhythmic hammering. After several blows, the metal goes back in the fire. One smith pulls out a bottle of beer, and they each take a swig. Then, out comes the glowing metal and the hammering resumes.',
  'genres': ['Short', 'Long'],
  'imdb': {'id': 5, 'rating': 6.2, 'votes': 1189},
  'lastupdated': '2015-08-26 00:03:50.133000000',
  'num_mflix_comments': 1,
  'plot': 'Three men hammer on an anvil and pass a bottle of beer around.',
  'rated': 'UNRATED',
  'released': datetime.datetime(1893, 5, 9, 0, 0),
  'runtime': 1,
  'title': 'Blacksmith Scene II',
  'to

Dari hasil tersebut dapat dilihat bahwa collection **movies** dan **clean_movies_faizah** ada document yang berbeda. Document diatas memiliki informasi filter **imdb** yang sama akan tetapi memiliki **judul** yang berbeda. Di **clean_movies_faizah** document dengan filter **{'imdb':{'id': 5, 'rating': 6.2, 'votes': 1189}}** memiliki **title** : **Blacksmith Scene** sedangkan pada **movies** collection dengan filter yang sama memiliki **title** : **Blacksmith Scene II** dan bersifat redundant karena ada 2 document dengan informasi sama di **movies** collection.

### 2. Get Keys/Field of movies and clean_movies_faizah

Validasi tentang apakah jumlah unique keys pada **movies** collection sudah sama dengan jumlah unique keys pada **clean_movie_faizah** collection.

**Get number of keys in 'movies' collection**

Sebelumnya saya telah mendefinisikan **unique keys** sebelum projecting. Sehingga untuk melihat jumlah fields dan isi fields nya sama dilihat dengan unique keys ini.

In [25]:
a

{'_id',
 'awards',
 'cast',
 'countries',
 'directors',
 'fullplot',
 'genres',
 'imdb',
 'languages',
 'lastupdated',
 'metacritic',
 'num_mflix_comments',
 'plot',
 'poster',
 'rated',
 'released',
 'runtime',
 'title',
 'tomatoes',
 'type',
 'writers',
 'year'}

In [21]:
len(a)

22

**Get number of keys in 'clean_movies_faizah' collection**

In [23]:
b = [list(doc.keys()) for doc in colmoviesi.find({})]
b = set([j for i in b for j in i])
b

{'_id',
 'awards',
 'cast',
 'countries',
 'directors',
 'fullPlot',
 'genres',
 'imdb',
 'languages',
 'lastupdated',
 'metacritic',
 'num_mflix_comments',
 'plot',
 'poster',
 'rated',
 'released',
 'runtime',
 'title',
 'tomatoes',
 'type',
 'writers',
 'year'}

In [24]:
len(b)

22

Dari hasil diatas, didapatkan informasi tentang jumlah **unique keys** dari **movies** dan **clean_movies_faizah** collection dan disimpulkan bahwa kedua collection tersebut telah memiliki jumlah **unique keys** yang sama yaitu 22.

### 3. Checking values in 'movies' same as in 'clean_movies_faizah'

Pada tahap ini akan dilakukan ***validasi*** apakah values yang terdapat di 'movies' collection sama dengan 'clean_movies_faizah'. ***validasi*** dilakukan dengan memilih satu document dengan filter **title**. 

Disini untuk checking values di document saya gunakan filter **title** : Blacksmith Scene yang saya temukan di **find_one** setelah projecting pada collection **clean_movies_faizah**.

In [74]:
#Showing document with filter 'title' in clean_movies_faizah
list(colmoviesi.find({'title': 'Blacksmith Scene'}))

[{'_id': ObjectId('5daffb1b986c4adedb754afa'),
  'awards': {'nominations': None, 'text': '1 win.', 'wins': '1'},
  'cast': ['Charles Kayser', 'John Ott'],
  'countries': ['USA'],
  'directors': ['William K.L. Dickson'],
  'fullPlot': 'A stationary camera looks at a large anvil with a blacksmith behind it and one on either side. The smith in the middle draws a heated metal rod from the fire, places it on the anvil, and all three begin a rhythmic hammering. After several blows, the metal goes back in the fire. One smith pulls out a bottle of beer, and they each take a swig. Then, out comes the glowing metal and the hammering resumes.',
  'genres': ['Short'],
  'imdb': {'id': 5, 'rating': 6.2, 'votes': 1189},
  'num_mflix_comments': 1,
  'plot': 'Three men hammer on an anvil and pass a bottle of beer around.',
  'rated': 'UNRATED',
  'released': 'datetime.datetime(1893, 05, 09, 0, 0)',
  'runtime': '1',
  'title': 'Blacksmith Scene',
  'tomatoes': {'lastUpdated': '', 'viewer': {'numReview

In [75]:
#Showing document with filter 'title' in movies
list(colmovies.find({'title': 'Blacksmith Scene'}))

[{'_id': ObjectId('573a1390f29313caabcd4135'),
  'awards': {'nominations': 0, 'text': '1 win.', 'wins': 1},
  'cast': ['Charles Kayser', 'John Ott'],
  'countries': ['USA'],
  'directors': ['William K.L. Dickson'],
  'fullplot': 'A stationary camera looks at a large anvil with a blacksmith behind it and one on either side. The smith in the middle draws a heated metal rod from the fire, places it on the anvil, and all three begin a rhythmic hammering. After several blows, the metal goes back in the fire. One smith pulls out a bottle of beer, and they each take a swig. Then, out comes the glowing metal and the hammering resumes.',
  'genres': ['Short'],
  'imdb': {'id': 5, 'rating': 6.2, 'votes': 1189},
  'lastupdated': '2015-08-26 00:03:50.133000000',
  'num_mflix_comments': 1,
  'plot': 'Three men hammer on an anvil and pass a bottle of beer around.',
  'rated': 'UNRATED',
  'released': datetime.datetime(1893, 5, 9, 0, 0),
  'runtime': 1,
  'title': 'Blacksmith Scene',
  'tomatoes': {'

Dari hasil diatas terlihat bahwa values  document pada **movies** collection dan **clean_movies_faizah** collection dengan filter **title: Blacksmith Scene** dihasilkan bahwa kedua collection tersebut memiliki values yang **sama**.

<a id='conclusions'></a>
## Conclusion

Diatas saya telah membuat collection baru dengan nama **clean_movies_faizah** dimana isinya berisi data dari **movies_initial** yang nanti hasilnya sama persis collection **movies** meliputi jumlah document, values document, jumlah unique keys, dan format document. 

Dari hasil analysis diatas dimana pembuatan collection baru dan validasi terdapat perbedaan jumlah document pada **clean_movies_faizah** dan **movies** yaitu terdapat 2 selisih document. Sedangkan untuk informasi yang lain sudah memenuhi validasi kesamaan kedua collection tersebut.

- Jumlah document pada collection **clean_movies_faizah** : 23539
- Jumlah document pada collection **movies** : 23541
- Jumlah **unique keys** : 22
- Nama **unique keys** :'_id', 'awards', 'cast', 'countries', 'directors', 'fullPlot', 'genres', 'imdb', 'languages', 'lastupdated', 'metacritic', 'num_mflix_comments', 'plot', 'poster', 'rated', 'released', 'runtime', 'title', 'tomatoes', 'type', 'writers','year'
- Values sama